services/payments/datastore.go

```diff
@@ -372,98 +184,129 @@ func newQLDBDatastore(ctx context.Context) (*qldbdriver.QLDBDriver, error) {
 	return driver, nil
 }
 
-const (
-	// StatePrepared - transaction prepared state
-	StatePrepared = "prepared"
-	// StateSubmitted - transaction prepared state
-	StateSubmitted = "submitted"
-)
+// insertPayment - perform a qldb insertion on a given transaction after doing all validation.
+// Returns the documentID for the record that was inserted.
+func (s Service) insertPayment(
+	ctx context.Context,
+	details PaymentDetails,
+) (string, error) {
+	unsignedPaymentState, err := UnsignedPaymentStateFromDetails(details, s.idempotencyNamespace)
+	if err != nil {
+		return "", err
+	}
+	_, signature, err := signPaymentState(
+		ctx,
+		s.kmsSigningClient,
+		s.kmsDecryptKeyArn,
+		*unsignedPaymentState,
+	)
+	if err != nil {
+		return "", err
+	}
 
-// InsertTransaction - perform a qldb insertion on the transactions.
-func (s Service) InsertTransaction(ctx context.Context, transaction *Transaction) (Transaction, error) {
-	enrichedTransaction, err := s.datastore.Execute(context.Background(), func(txn qldbdriver.Transaction) (interface{}, error) {
-		// for all of the transactions load up a check to see if this transaction has already existed
-		// or not, then perform the insertion of the records.
-		resp := Transaction{}
+	paymentState := PaymentState{
+		UnsafePaymentState: unsignedPaymentState.UnsafePaymentState,
+		ID: unsignedPaymentState.ID,
+		Signature: []byte(signature),
```

Comment:
why do we need to reconstruct the object? can't we just assign the signature?

Reply:
It was because of naming confusion but I went ahead and changed from `unsignedPaymentState` to `paymentStateForSigning` and removed the recreation of the object. That should prevent confusion here.